src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameHeaders.cs

```diff
@@ -15,9 +15,10 @@ public abstract class FrameHeaders : IHeaderDictionary
     {
         protected bool _isReadOnly;
         protected Dictionary<string, StringValues> MaybeUnknown;
-
         protected Dictionary<string, StringValues> Unknown => MaybeUnknown ?? (MaybeUnknown = new Dictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase));
 
+        public long? ContentLength { get; set; }
```

Comment:
We should have a custom setter, so we can throw immediately if this is set to a negative value or _isReadOnly is true.